File: src/clawcu/a2a/sidecar_plugin/openclaw/sidecar/outbound_limit.js
"use strict";

// Self-origin outbound rate limit shared by /a2a/outbound and /mcp
// tool-call handlers (a2a-design-4.md §P1-B). Hop budget caps depth
// (A→B→A→B runaway); this caps breadth (one LLM turn fires 200 parallel
// a2a_call_peer calls and nukes the provider quota).
//
// Key: thread_id when present, else the caller's own registered name
// (`self:<name>`). Limit: N calls / rolling 60s / key. Default 60/min,
// tunable via A2A_OUTBOUND_RATE_LIMIT env var.

const DEFAULT_RPM = 60;
const WINDOW_MS = 60_000;

function readRpm(env) {
  const raw = (env || process.env).A2A_OUTBOUND_RATE_LIMIT;
  if (raw === undefined || raw === null || String(raw).trim() === "") return DEFAULT_RPM;
  const n = Number(raw);
  if (!Number.isFinite(n) || n <= 0 || Math.floor(n) !== n) return DEFAULT_RPM;
  return n;
}

function keyFor({ threadId, selfName }) {
  if (typeof threadId === "string" && threadId) return `thread:${threadId}`;
  return `self:${selfName || "anon"}`;
}

function createOutboundLimiter({ rpm, nowFn = Date.now } = {}) {
  const limit = Number.isFinite(rpm) && rpm > 0 ? rpm : DEFAULT_RPM;
  const hits = new Map();

  function check(key) {
    const now = nowFn();
    const cutoff = now - WINDOW_MS;
    const arr = hits.get(key) || [];
    // prune old entries
    while (arr.length && arr[0] <= cutoff) arr.shift();
    if (arr.length >= limit) {
      const retryAfterMs = arr[0] + WINDOW_MS - now;
      hits.set(key, arr);
      return { allowed: false, retryAfterMs: Math.max(0, retryAfterMs), limit };
    }
    arr.push(now);
    hits.set(key, arr);
    return { allowed: true, count: arr.length, limit };
  }

  // P1-J (a2a-design-5.md): sweep buckets whose deques have emptied out
  // past the window. Called opportunistically — cheap when there are few
  // keys, fine to skip under load.
  function sweep() {
    const now = nowFn();
    const cutoff = now - WINDOW_MS;
    for (const [k, arr] of hits) {
      while (arr.length && arr[0] <= cutoff) arr.shift();
      if (arr.length === 0) hits.delete(k);
    }
  }

  function size() {
    return hits.size;
  }

  function reset() {
    hits.clear();
  }

  return { check, reset, sweep, size, limit };
}

const DEFAULT_SWEEP_INTERVAL_MS = 300_000;

function readSweepIntervalMs(env) {
  const raw = (env || process.env).A2A_OUTBOUND_SWEEP_INTERVAL_MS;
  if (raw === undefined || raw === null || String(raw).trim() === "") return DEFAULT_SWEEP_INTERVAL_MS;
  const n = Number(raw);
  if (!Number.isFinite(n) || Math.floor(n) !== n) return DEFAULT_SWEEP_INTERVAL_MS;
  return Math.max(0, n);
}

// a2a-design-6.md §P2-L: wire a periodic sweep so long-running sidecars
// with many distinct thread_ids over days don't accumulate empty buckets.
// `scheduler` is injectable for tests; defaults to setInterval. Returns
// the handle (or null when intervalMs <= 0). The caller should .unref()
// the handle so the timer doesn't block graceful shutdown.
function createSweepTimer({ limiter, intervalMs, scheduler = setInterval }) {
  if (!limiter || typeof limiter.sweep !== "function") return null;
  if (!Number.isFinite(intervalMs) || intervalMs <= 0) return null;
  const handle = scheduler(() => {
    try {
      limiter.sweep();
    } catch (err) {
      // a2a-design-7.md §P2-N: sweep is opportunistic cleanup, never
      // load-bearing — still swallow, but leave a breadcrumb so an
      // operator grepping logs can see something went wrong.
      try {
        console.warn(
          `[sidecar] outbound-sweep failed: ${err && err.message ? err.message : err}`,
        );
      } catch (_logErr) {
        // console itself faked/broken — keep the timer alive regardless.
      }
    }
  }, intervalMs);
  if (handle && typeof handle.unref === "function") handle.unref();
  return handle;
}

module.exports = {
  DEFAULT_RPM,
  WINDOW_MS,
  DEFAULT_SWEEP_INTERVAL_MS,
  readRpm,
  readSweepIntervalMs,
  keyFor,
  createOutboundLimiter,
  createSweepTimer,
};

File: src/clawcu/a2a/sidecar_plugin/openclaw/sidecar/outbound_limit.py
"""Self-origin outbound rate limiter (Python port of outbound_limit.js).

Shared limiter for /a2a/outbound and MCP tools/call handlers. The key is
either `thread:<threadId>` or `self:<selfName>` (defaulting to "anon").
Default 60 calls / 60s, configurable via A2A_OUTBOUND_RATE_LIMIT.
"""
from __future__ import annotations

import os
import threading
import time
from collections import deque
from dataclasses import dataclass
from typing import Callable, Deque, Dict, Mapping, Optional

DEFAULT_RPM = 60
WINDOW_MS = 60_000
DEFAULT_SWEEP_INTERVAL_MS = 300_000


def _default_now_ms() -> int:
    return int(time.time() * 1000)


def _coerce_int_env(raw, fallback: int, allow_zero: bool = False, clamp_nonneg: bool = False) -> int:
    if raw is None:
        return fallback
    s = str(raw).strip()
    if s == "":
        return fallback
    try:
        n = float(s)
    except (TypeError, ValueError):
        return fallback
    if n != int(n):
        return fallback
    n_int = int(n)
    if clamp_nonneg:
        return max(0, n_int)
    if n_int <= 0 and not allow_zero:
        return fallback
    return n_int


def read_rpm(env: Optional[Mapping[str, str]] = None) -> int:
    e = env if env is not None else os.environ
    return _coerce_int_env(e.get("A2A_OUTBOUND_RATE_LIMIT"), DEFAULT_RPM)


def read_sweep_interval_ms(env: Optional[Mapping[str, str]] = None) -> int:
    e = env if env is not None else os.environ
    return _coerce_int_env(
        e.get("A2A_OUTBOUND_SWEEP_INTERVAL_MS"),
        DEFAULT_SWEEP_INTERVAL_MS,
        clamp_nonneg=True,
    )


def key_for(thread_id: Optional[str] = None, self_name: Optional[str] = None) -> str:
    if isinstance(thread_id, str) and thread_id:
        return f"thread:{thread_id}"
    return f"self:{self_name or 'anon'}"


@dataclass
class OutboundDecision:
    allowed: bool
    limit: int
    count: int = 0
    retry_after_ms: int = 0


class OutboundLimiter:
    def __init__(
        self,
        rpm: Optional[int] = None,
        now_fn: Callable[[], int] = _default_now_ms,
    ) -> None:
        self.limit = rpm if (isinstance(rpm, int) and rpm > 0) else DEFAULT_RPM
        self.now_fn = now_fn
        self._hits: Dict[str, Deque[int]] = {}

    def check(self, key: str) -> OutboundDecision:
        now = self.now_fn()
        cutoff = now - WINDOW_MS
        arr = self._hits.get(key)
        if arr is None:
            arr = deque()
            self._hits[key] = arr
        while arr and arr[0] <= cutoff:
            arr.popleft()
        if len(arr) >= self.limit:
            retry_after_ms = max(0, arr[0] + WINDOW_MS - now)
            return OutboundDecision(allowed=False, limit=self.limit, retry_after_ms=retry_after_ms)
        arr.append(now)
        return OutboundDecision(allowed=True, limit=self.limit, count=len(arr))

    def sweep(self) -> None:
        now = self.now_fn()
        cutoff = now - WINDOW_MS
        empty_keys = []
        for k, arr in self._hits.items():
            while arr and arr[0] <= cutoff:
                arr.popleft()
            if not arr:
                empty_keys.append(k)
        for k in empty_keys:
            self._hits.pop(k, None)

    def size(self) -> int:
        return len(self._hits)

    def reset(self) -> None:
        self._hits.clear()


def create_outbound_limiter(
    rpm: Optional[int] = None,
    now_fn: Callable[[], int] = _default_now_ms,
) -> OutboundLimiter:
    return OutboundLimiter(rpm=rpm, now_fn=now_fn)


class SweepTimer:
    """Best-effort periodic sweep. `.unref()`-style: daemon thread, so it
    doesn't block interpreter shutdown.
    """

    def __init__(
        self,
        limiter: OutboundLimiter,
        interval_ms: int,
        logger=None,
    ) -> None:
        self.limiter = limiter
        self.interval_s = interval_ms / 1000.0
        self._stop = threading.Event()
        self._thread: Optional[threading.Thread] = None
        self._logger = logger

    def start(self) -> None:
        if self._thread is not None:
            return
        t = threading.Thread(target=self._run, name="a2a-outbound-sweep", daemon=True)
        self._thread = t
        t.start()

    def stop(self) -> None:
        self._stop.set()
        t = self._thread
        self._thread = None
        if t is not None:
            t.join(timeout=1.0)

    def _run(self) -> None:
        while not self._stop.wait(self.interval_s):
            try:
                self.limiter.sweep()
            except Exception as err:  # pragma: no cover - defensive
                if self._logger is not None:
                    try:
                        self._logger.warn(f"[sidecar] outbound-sweep failed: {err}")
                    except Exception:
                        pass


def create_sweep_timer(
    limiter: OutboundLimiter,
    interval_ms: int,
    logger=None,
) -> Optional[SweepTimer]:
    if limiter is None or not hasattr(limiter, "sweep"):
        return None
    if not isinstance(interval_ms, int) or interval_ms <= 0:
        return None
    timer = SweepTimer(limiter=limiter, interval_ms=interval_ms, logger=logger)
    timer.start()
    return timer
